grafonnet-base/helpers.libsonnet

```diff
@@ -1,11 +1,137 @@
+local d = import 'github.com/jsonnet-libs/docsonnet/doc-util/main.libsonnet';
+local xtd = import 'github.com/jsonnet-libs/xtd/main.libsonnet';
+
 {
-  getAttributes(attributes, origin):
-    {
-      [n]: origin[n]
-      for n in attributes
-    }
-    + {
-      ['#' + n]: origin['#' + n]
-      for n in attributes
-    },
+  local root = self,
+
+  last(arr): std.reverse(arr)[0],
+  allButLast(arr): arr[0:std.length(arr) - 1],
+
+  getContent(source, path):
+    local content = xtd.jsonpath.getJSONPath(source, path);
+    if content == null
+    then self.last(xtd.string.splitEscape(path, '.'))
+    else content,
+
+  setContent(content, path):
+    std.foldr(
+      function(k, acc)
+        { [k]+: acc },
+      xtd.string.splitEscape(path, '.'),
+      content
+    ),
+
+  hideContent(source, path):
+    local splitPath = xtd.string.splitEscape(path, '.');
+    local content = root.getContent(source, path);
+    std.foldr(
+      function(k, acc)
+        { [k]+: acc },
+      root.allButLast(splitPath),
+      { [root.last(splitPath)]:: content }
+    ),
+
+  removeContent(source, path):
+    local splitPath = xtd.string.splitEscape(path, '.');
+    std.foldr(
+      function(k, acc)
+        { [k]+: acc },
+      root.allButLast(splitPath),
+      { [root.last(splitPath)]:: {} }
+    ),
+
+  transform(source, from, to):
+    local content = root.getContent(source, from);
+    root.setContent(content, to)
+  ,
+
+  groupingsToTransformArray(groupings, keyPrefix='', keySuffix='', separator='.'):
```

Comment:
comments would be helpful to make it clear what transforms these functions offer

Reply:
It would help me review, which is why I asked.

Reply:
comments added